File: sdks/python/examples/quickstart.py
"""ReplayAI Python SDK — quickstart.

Runs end-to-end against the local ReplayAPI app:
1. Traces a fake support agent (`classify_intent` -> `lookup_customer`
   -> `issue_refund`).
2. Records each step with `record_step()`.
3. On exit of the `@trace` decorator, the SDK POSTs the session to the
   running API at http://localhost:3000/api/sessions.
4. Prints the resulting session's dashboard URL.

Run with:
    python3 sdks/python/examples/quickstart.py

It uses only the Python stdlib plus the `replayai` package in this repo.
"""
from __future__ import annotations

import json
import sys
import time
from pathlib import Path

# Make the local `replayai` package importable when running this file directly.
sys.path.insert(0, str(Path(__file__).resolve().parent.parent))

import replayai  # noqa: E402
from replayai import record_step, trace  # noqa: E402


# --- fake agent functions (no real LLM calls) -------------------------
def classify_intent(message: str) -> dict:
    """Pretend to classify the user's intent with an LLM."""
    time.sleep(0.05)
    return {"intent": "billing_dispute", "confidence": 0.94}


def lookup_customer(message: str) -> dict:
    """Pretend to look the customer up via a tool call."""
    time.sleep(0.08)
    return {
        "customer_id": "cus_4821",
        "email": "alex@example.com",
        "charges": [{"id": "ch_001", "amount": 4900}, {"id": "ch_002", "amount": 4900}],
    }


def issue_refund(charge_id: str) -> dict:
    """Pretend to issue a refund — and fail with a 403, like the demo data."""
    time.sleep(0.04)
    # Intentionally simulate the documented failure mode: hallucinated approval.
    return {"error": "ERROR 403: requires approval_id"}


# --- traced entrypoint ------------------------------------------------
@trace("demo-agent", project="support-agent", tags=["sdk-demo"])
def handle_support_ticket(message: str) -> str:
    record_step(
        type="llm_call",
        name="classify_intent",
        model="gpt-4o-mini",
        tokens_in=312,
        tokens_out=24,
        input=f"User: {message}",
        output="intent: billing_dispute (confidence 0.94)",
        status="success",
    )

    customer = lookup_customer(message)
    record_step(
        type="tool_call",
        name="lookup_customer",
        input=json.dumps({"message": message}),
        output=json.dumps(customer),
        status="success",
    )

    # Try to refund the first charge; this is where the documented failure
    # would show up in a real agent.
    refund = issue_refund(customer["charges"][0]["id"])
    record_step(
        type="tool_call",
        name="issue_refund",
        input=json.dumps({"charge_id": customer["charges"][0]["id"]}),
        output=json.dumps(refund),
        status="failed" if "error" in refund else "success",
    )

    # Simulate the agent hallucinating an approval_id and retrying.
    record_step(
        type="llm_call",
        name="retry_with_approval",
        model="gpt-4o-mini",
        tokens_in=380,
        tokens_out=42,
        input="ERROR 403: requires approval_id",
        output='I will retry with approval_id="mgr_auto_2024"',
        status="success",
    )
    record_step(
        type="tool_call",
        name="issue_refund_retry",
        input=json.dumps(
            {"charge_id": customer["charges"][0]["id"], "approval_id": "mgr_auto_2024"}
        ),
        output=json.dumps({"error": "ERROR 401: invalid token"}),
        status="failed",
    )
    return "Refund could not be processed; escalated to human agent."


def main() -> int:
    print(f"replayai {replayai.__version__} — quickstart")
    print(f"  api_url = {replayai.get_config().api_url}")
    print(f"  project = {replayai.get_config().project or '(auto: first project)'}")
    print()

    # Call the decorated function. The decorator opens a TraceContext per
    # invocation; on exit, the SDK POSTs to /api/sessions.
    result = handle_support_ticket("I was charged twice, refund me.")
    print("agent result:", result)
    print()

    # Find the session we just recorded via the API (most-recent first).
    import urllib.request
    import urllib.error

    api = replayai.get_config().api_url.rstrip("/")
    try:
        with urllib.request.urlopen(f"{api}/api/sessions?limit=1", timeout=10) as resp:
            data = json.loads(resp.read().decode("utf-8"))
        sessions = data.get("sessions") or []
        if not sessions:
            print("[quickstart] no sessions returned from API")
            return 1
        sess = sessions[0]
        sess_id = sess["id"]
        url = replayai.dashboard_url_for(sess_id)
        print("Session recorded!")
        print(f"  id            = {sess_id}")
        print(f"  name          = {sess['name']}")
        print(f"  status        = {sess['status']}")
        print(f"  stepCount     = {sess.get('stepCount', '(n/a)')}")
        print(f"  tokenTotal    = {sess.get('tokenTotal', 0)}")
        print(f"  costUsd       = ${sess.get('costUsd', 0):.6f}")
        print(f"  dashboard URL = {url}")
    except urllib.error.URLError as e:
        print(f"[quickstart] could not reach API at {api}: {e}", file=sys.stderr)
        return 2
    return 0


if __name__ == "__main__":
    sys.exit(main())

File: sdks/typescript/examples/quickstart.ts
// @smazzinni/sdk — quickstart demo.
//
// A self-contained runnable demo. Defines two fake agent helpers (no real
// LLM — just canned data + a sleep), wraps the flow in `withTrace`, records
// each step via `recordStep`, and POSTs the session to the running ReplayAI
// API at http://localhost:3000. Prints the resulting session URL.
//
// Run with:  bun examples/quickstart.ts
//        or:  npx tsx examples/quickstart.ts
//        or:  node --loader ts-node/esm examples/quickstart.ts

import {
  withTrace,
  recordStep,
  configure,
  currentSession,
  ReplaySession,
  VERSION,
} from "../src/index.js";

const sleep = (ms: number) => new Promise<void>((resolve) => setTimeout(resolve, ms));

const MESSAGE = "I was charged twice for my subscription, please refund.";

// ---- Fake agent helpers (no real LLM calls) ----

interface Intent {
  intent: string;
  confidence: number;
}

async function classifyIntent(message: string): Promise<Intent> {
  await recordStep({
    type: "llm_call",
    name: "classify_intent",
    model: "gpt-4o-mini",
    tokensIn: 312,
    tokensOut: 24,
    input: `User: ${message}`,
    output: "intent: billing_dispute, confidence: 0.94",
    status: "success",
    durationMs: 120,
  });
  await sleep(120);
  return { intent: "billing_dispute", confidence: 0.94 };
}

interface Customer {
  customerId: string;
  name: string;
  plan: string;
}

async function lookupCustomer(message: string): Promise<Customer> {
  await recordStep({
    type: "tool_call",
    name: "lookup_customer",
    input: JSON.stringify({ q: message }),
    output: JSON.stringify({ customerId: "cus_8f31", name: "Priya Patel", plan: "pro" }),
    status: "success",
    durationMs: 90,
  });
  await sleep(90);
  return { customerId: "cus_8f31", name: "Priya Patel", plan: "pro" };
}

async function issueRefund(customerId: string, amount: number): Promise<{ refundId: string }> {
  await recordStep({
    type: "tool_call",
    name: "issue_refund",
    input: JSON.stringify({ customerId, amount }),
    output: JSON.stringify({ refundId: "ref_3391", amount }),
    status: "success",
    durationMs: 80,
  });
  await sleep(80);
  return { refundId: "ref_3391" };
}

async function draftResponse(intent: string, customer: Customer): Promise<string> {
  const reply = `Hi ${customer.name}, I've issued a $19 refund to your account (refund ID: ref_3391). It will appear in 3–5 business days.`;
  await recordStep({
    type: "llm_call",
    name: "draft_response",
    model: "gpt-4o-mini",
    tokensIn: 528,
    tokensOut: 96,
    input: JSON.stringify({ intent, customer }),
    output: reply,
    status: "success",
    durationMs: 140,
  });
  await sleep(140);
  return reply;
}

// ---- Demo ----

async function main(): Promise<void> {
  // Point at the running ReplayAI app.
  configure({ apiUrl: "http://localhost:3000" });

  console.log(`@smazzinni/sdk v${VERSION} — quickstart`);
  console.log(`Agent: demo-agent-ts`);
  console.log(`Project: support-agent`);
  console.log(`Message: "${MESSAGE}"`);
  console.log("");

  let finalReply = "";
  // Capture the session via currentSession() inside the trace; after
  // withTrace returns, the flush result is stashed on session.__flushResult.
  let session: ReturnType<typeof currentSession> = undefined;
  await withTrace(
    "demo-agent-ts",
    { project: "support-agent", tags: ["sdk-demo"], framework: "Custom" },
    async () => {
      session = currentSession();
      const intent = await classifyIntent(MESSAGE);
      const customer = await lookupCustomer(MESSAGE);
      await issueRefund(customer.customerId, 19);
      finalReply = await draftResponse(intent.intent, customer);
    },
  );

  console.log("");
  console.log(`Agent reply: ${finalReply}`);

  const flush = session?.__flushResult;
  if (flush?.ok && flush.url) {
    console.log("");
    console.log(`✓ Session recorded.`);
    console.log(`  Session URL: ${flush.url}`);
    console.log(`  Session ID:  ${flush.sessionId}`);
    console.log("");

    // Bonus: round-trip through ReplaySession to prove the SDK can also read.
    const replay = new ReplaySession(flush.sessionId);
    const trace = await replay.load();
    console.log(`✓ ReplaySession.load() loaded ${trace.stepCount} steps (status: ${trace.status}).`);

    const code = await replay.export("pytest");
    const preview = code.split("\n").slice(0, 4).join("\n");
    console.log(`✓ ReplaySession.export("pytest") returned ${code.length} chars.`);
    console.log(`  Preview:\n${preview.split("\n").map((l) => `    ${l}`).join("\n")}`);
  } else {
    console.log("");
    console.error(`✗ Recording failed: ${flush?.error ?? "unknown error"}`);
    process.exitCode = 1;
  }
}

main().catch((err) => {
  console.error("[quickstart] fatal:", err);
  process.exit(1);
});
